apps/emqx_ds_builtin_raft/test/emqx_ds_builtin_raft_SUITE.erl

```diff
@@ -250,121 +250,119 @@ t_rebalance(Config) ->
-                %% Now join the rest of the sites:
-                {N2, assign_db_sites, Sites}
-            ],
-            Stream1 = emqx_utils_stream:interleave(
-                [
-                    {20, Stream0},
-                    emqx_utils_stream:const(add_generation)
-                ],
-                false
-            ),
-            Stream = emqx_utils_stream:interleave(
-                [
-                    {50, Stream1},
-                    emqx_utils_stream:list(Sequence)
-                ],
-                true
-            ),
+    Sequence = [
+        %% Join the second site to the DB replication sites:
+        {N1, join_db_site, S2},
+        %% Should be a no-op:
+        {N2, join_db_site, S2},
+        %% Now join the rest of the sites:
+        {N2, assign_db_sites, Sites}
+    ],
+    Stream1 = emqx_utils_stream:interleave(
+        [
+            {20, Stream0},
+            emqx_utils_stream:const(add_generation)
+        ],
+        false
+    ),
+    Stream = emqx_utils_stream:interleave(
+        [
+            {50, Stream1},
+            emqx_utils_stream:list(Sequence)
+        ],
+        true
+    ),
 
-            %% 1.2 Verify that all nodes have the same view of metadata storage:
-            [
-                ?defer_assert(
-                    ?assertEqual(
-                        [S1],
-                        ?ON(Node, emqx_ds_builtin_raft_meta:db_sites(?DB)),
-                        #{
-                            msg => "Initially, only S1 should be responsible for all shards",
-                            node => Node
-                        }
-                    )
-                )
-             || Node <- Nodes
-            ],
+    %% 1.2 Verify that all nodes have the same view of metadata storage:
+    [
+        ?defer_assert(
+            ?assertEqual(
+                [S1],
+                ?ON(Node, emqx_ds_builtin_raft_meta:db_sites(?DB)),
+                #{
+                    msg => "Initially, only S1 should be responsible for all shards",
+                    node => Node
+                }
+            )
+        )
+     || Node <- Nodes
+    ],
 
-            %% 2. Start filling the storage:
-            emqx_ds_raft_test_helpers:apply_stream(?DB, Nodes, Stream),
-            timer:sleep(5000),
-            emqx_ds_raft_test_helpers:verify_stream_effects(
-                ?DB, ?FUNCTION_NAME, Nodes, TopicStreams
-            ),
-            [
-                ?defer_assert(
-                    ?assertEqual(
-                        16 * 3 div length(Nodes),
-                        n_shards_online(Node, ?DB),
-                        "Each node is now responsible for 3/4 of the shards"
-                    )
-                )
-             || Node <- Nodes
-            ],
+    %% 2. Start filling the storage:
+    emqx_ds_raft_test_helpers:apply_stream(?DB, Nodes, Stream),
+    timer:sleep(5000),
+    emqx_ds_raft_test_helpers:verify_stream_effects(
```

Comment:
The trouble is `?defer_assert` that is used extensively throughout the test case (in `verify_stream_effects/4` also) expects that we're inside snabbkaffe two-phase test run.